File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> response;
        preorder(root, response, 0);
        return response;
    }

    void preorder(TreeNode* root, vector<vector<int>>& response, int height) {
        if (root == nullptr) {
            return;
        }
        if (response.size() == height) {
            response.emplace_back();
        }

        response[height].push_back(root->val);
        preorder(root->left, response, height + 1);
        preorder(root->right, response, height + 1);
    }
};

File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    // Time complexity O(N) where N is number of nodes in tree and space O(N)
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> response = new ArrayList<>();
        preOrder(root, 0, response);
        return response;
    }
    
    public static void preOrder(TreeNode root, int height, List<List<Integer>> res) {
        if(root == null)
            return;

        if(res.size() == height)
            res.add(new ArrayList<>());

        res.get(height).add(root.val);

        preOrder(root.left, height + 1, res);
        preOrder(root.right, height + 1, res);
    }
}